rusoto/core/src/reactor.rs

```diff
@@ -0,0 +1,225 @@
+//! Default implementations for `DispatchSignedRequest` and `ProvideAwsCredentials` backed by
+//! an implicit event loop.
+//!
+//! The purpose of this module is to enable the use of Rusoto without requiring any knowledge about
+//! `tokio`.
+//!
+//! It does this by providing two simple structs, `CredentialsProvider` and `RequestDispatcher`,
+//! which can be plugged into any AWS client exposed by Rusoto.
+//!
+//! These implementations are backed by an implicit event loop running on a background thread, so no
+//! manual event loop setup is required.
+//!
+//! The implicit event loop is crated lazily on first use of these implementations, meaning that advanced
+//! users who wish to interface more directly with tokio can do so by using the otherwise provided
+//! implementations without causing any extra background threads or event loops to be spawned.
+
+use std::io::{Result as IoResult};
+use std::rc::Rc;
+use std::thread;
+
+use futures::{Async, Future, Poll, Stream};
+use futures::future::{Either, ok};
+use futures::sync::{mpsc, oneshot};
+use tokio_core::reactor::{Core, Handle, Remote};
+
+use super::{
+    SignedRequest, HttpResponse, HttpDispatchError, DispatchSignedRequest,
+    AwsCredentials, CredentialsError, ProvideAwsCredentials,
+    default_tls_client, TlsError, DefaultCredentialsProvider
+};
+
+lazy_static! {
+    static ref DEFAULT_REACTOR: Reactor = {
```

Comment:
Is having a single `Reactor` a standard pattern?  My understanding of event loops says yes but I'd like to make sure. 😄

Reply:
I've seen it this way in other languages, but I have to say I don't know if it is for tokio. I'd also be interested in knowing this.

Reply:
So, answering this requires going into a bit of context, so here it goes :)

One of my main goals when approaching this feature was to make rusoto still usable in a non-async codebase without a lot of extra ceremony required: Effectively the migration path to the new rusoto version should be as easy as possible for existing consumers.

In order to achieve that, I added the `RusotoFuture::sync` method as a simple way of blocking the current thread until a response arrives. However, calling `Futures::wait` (which `::sync` does internally) is not enough to make it work. Without an event loop in the background driving the I/O, `::sync` would just block forever.

While a single "default" event loop is not a standard pattern for async Rust at the moment (the hyper client for instance needs to be instantiated by referencing an existing event loop), and we should still support the more "advanced" use-cases (which this PR does), I believe introducing a default event loop here is a good idea:

Tokio is currently planning a reform of the tokio crates, with a very similar goal, to simplify the integration story for the majority of tokio users. You can read more what they are planning [here](https://tokio.rs/blog/tokio-reform/). As part of this effort, they are going to introduce a default event loop, very similar to what the reactor is providing here.

Unfortunately, the tokio reform still seems to be multiple months (~6?) out at this point. So while we can't take advantage of the new tokio default event loop just yet, the way the default loop is implemented in this PR will allow us to migrate to the new tokio pattern as soon as it is released, without any breaking changes.

So tl;dr: There is no standard pattern for this yet, however tokio is moving into this direction and the implementation in this PR aligns us well with where tokio is headed.